src/main/java/hudson/remoting/Engine.java

```diff
@@ -234,54 +229,41 @@ public void run() {
                     return;
                 }
 
-                Socket s = connect(port);
-
                 events.status("Handshaking");
-
-                DataOutputStream dos = new DataOutputStream(s.getOutputStream());
-                BufferedInputStream in = new BufferedInputStream(s.getInputStream());
-
-                dos.writeUTF("Protocol:JNLP2-connect");
-                Properties props = new Properties();
-                props.put("Secret-Key", secretKey);
-                props.put("Node-Name", slaveName);
-                if (cookie!=null)
-                    props.put("Cookie", cookie);
-                ByteArrayOutputStream o = new ByteArrayOutputStream();
-                props.store(o, null);
-                dos.writeUTF(o.toString("UTF-8"));
-
-                String greeting = readLine(in);
-                if (greeting.startsWith("Unknown protocol")) {
-                    LOGGER.info("The server didn't understand the v2 handshake. Falling back to v1 handshake");
-                    s.close();
-                    s = connect(port);
-                    in = new BufferedInputStream(s.getInputStream());
-                    dos = new DataOutputStream(s.getOutputStream());
-
-                    dos.writeUTF("Protocol:JNLP-connect");
-                    dos.writeUTF(secretKey);
-                    dos.writeUTF(slaveName);
-
-                    greeting = readLine(in); // why, oh why didn't I use DataOutputStream when writing to the network?
-                    if (!greeting.equals(GREETING_SUCCESS)) {
-                        onConnectionRejected(greeting);
-                        continue;
-                    }
-                } else {
-                    if (greeting.equals(GREETING_SUCCESS)) {
-                        Properties responses = readResponseHeaders(in);
-                        cookie = responses.getProperty("Cookie");
-                    } else {
-                        onConnectionRejected(greeting);
-                        continue;
+                Socket jnlpSocket = connect(port);
+                DataOutputStream outputStream = new DataOutputStream(jnlpSocket.getOutputStream());
+                BufferedInputStream inputStream = new BufferedInputStream(jnlpSocket.getInputStream());
+                String response = "";
```

Comment:
Possibly more readable to move this variable inside the loop, and have a `boolean` flag for whether the handshake succeeded.